embassy-microchip/src/i2c.rs

```diff
@@ -0,0 +1,1183 @@
+pub enum AbortReason {
+    /// A bus operation was not acknowledged.
+    NoAcknowledge,
+
+    /// Lost arbitration
+    ArbitrationLoss,
+
+    /// Other reason
+    Other,
+}
+
+/// I2C error bus timeout reason
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum BusTimeoutReason {
+    /// Clock high, data high timeout
+    ClockHighDataHigh,
+
+    /// Clock high, data low timeout
+    ClockHighDataLow,
+
+    /// Slave cumulative timeout
+    SlaveCumulativeTimeout,
+
+    /// Master cumulative timeout
+    MasterCumulativeTimeout,
+
+    /// Device timeout
+    DeviceTimeout,
+}
+
+/// I2C error
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum Error {
+    /// I2C abort with error
+    Abort(AbortReason),
+
+    /// Bus timeout
+    BusTimeout(BusTimeoutReason),
+
+    /// User passed in a read buffer with 0 length
+    InvalidReadBufferLength,
+
+    /// User passed in a write buffer with 0 length
+    InvalidWriteBufferLength,
+
+    /// Target I2C address is out of range
+    AddressOutOfRange(u8),
+
+    /// Target I2C address is reserved
+    AddressReserved(u8),
+}
+
+/// I2C config error
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum ConfigError {
+    /// Max I2C frequency is 1MHz
+    FrequencyTooHigh,
+
+    /// The baud rate clock is too slow to support the requested
+    /// frequency
+    ClockTooSlow,
+
+    /// The baud rate clock is too fast to support the requested
+    /// frequency
+    ClockTooFast,
+}
+
+/// I2C bus speeds
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum BusSpeed {
+    /// 100kHz
+    Standard,
+
+    /// 400kHz,
+    Fast,
+
+    /// 1MHz,
+    FastPlus,
+}
+
+/// I2C config
+#[non_exhaustive]
+#[derive(Copy, Clone)]
+pub struct Config {
+    /// Bus speed.
+    pub speed: BusSpeed,
+
+    /// Own address 1 in 7-bit format.
+    ///
+    /// The internal master should not attempt transactions to a slave
+    /// with the same address as `addr1`. This represents an illegal
+    /// operation.
+    ///
+    /// Passing `None` results in the value 0 being written to the Own
+    /// Address register and General Call Match to be disabled.
+    pub addr1: Option<u8>,
+
+    /// Own address 2 in 7-bit format.
+    ///
+    /// The internal master should not attempt transactions to a slave
+    /// with the same address as `addr2`. This represents an illegal
+    /// operation.
+    ///
+    /// Passing `None` results in the value 0 being written to the Own
+    /// Address register and General Call Match to be disabled.
+    pub addr2: Option<u8>,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            speed: BusSpeed::Standard,
+            addr1: Some(0x7f),
+            addr2: Some(0x7f),
```

Comment:
Would `None` be a better default, and set `gc_dis=1` by default? (see below). I guess not too important now given multi-master isn't supported.

Reply:
done